### PR TITLE
Import the existing bioblend tutorial, exercises 

### PR DESCRIPTION
**Pull request by hexylena**
Converted from https://github.com/nsoranzo/bioblend-tutorial via jupytext and basically directly into the GTN with a little bit of extra formatting for questions to be prettier.

Tutorial | Notebook
--- | ---
image | image

### CODE REVIEW

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+>
+> 1. TOC
+> {:toc}
+>
+{: .agenda}
+
+
+# Interacting with histories in Galaxy API
+
+We are going to use the [requests](http://python-requests.org/) Python library to communicate via HTTP with the Galaxy server. To start, let's define the connection parameters.
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:49]:
```suggestion
We are going to use the [requests](https://requests.readthedocs.io/) Python library to communicate via HTTP with the Galaxy server. To start, let's define the connection parameters.
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,806 @@
...
+3. Click on "Manage API key"
+4. Generate an API key if needed, then copy the alphanumeric string and paste it as the value of the `api_key` variable below.
+
+```python
+from pprint import pprint
+
+import json
+
+import requests
+from six.moves.urllib.parse import urljoin
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:63]:
```suggestion
import json
from pprint import pprint
from urllib.parse import urljoin

import requests
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,806 @@
...
+> ### {% icon question %} Question: Initialise
+> First, define the connection parameters. What variables do you need?
+> > ### {% icon solution %} Solution
+> > ```python
+> > from pprint import pprint
+> >
+> > import json
+> >
+> > import requests
+> > from six.moves.urllib.parse import urljoin
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:185]:
```suggestion
> > import json
> > from pprint import pprint
> > from urllib.parse import urljoin
> >
> > import requests
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,806 @@
...
+- `creating job`: id of the job which created this dataset
+- `download_url`: URL to download the dataset
+- `file_ext`: the Galaxy data type of this dataset
+- `file_size`
+- `genome_build`: the genome build (dbkey) associated to this dataset.
+
+**New resources** are created with POST requests. The uploaded **data needs to be serialized** in a JSON string. For example, to create a new history:
+
+```python
+params = {'key': api_key}
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:142]:
```suggestion
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,806 @@
...
+data = {'name': 'New history'}
+r = requests.post(base_url + '/histories', data=json.dumps(data), params=params, headers={'Content-Type': 'application/json'})
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:144]:
```suggestion
r = requests.post(base_url + "/histories", data=json.dumps(data), headers=headers)
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,806 @@
...
+new_hist = r.json()
+pprint(new_hist)
+```
+
+The return value of a POST request is a dictionary with detailed info about the created resource.
+
+To **update** a resource, make a PUT request, e.g. to change the history name:
+
+```python
+params = {'key': api_key}
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:154]:
```suggestion
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,806 @@
...
+data = {'name': 'Updated history'}
+r = requests.put(base_url + '/histories/' + new_hist['id'], json.dumps(data), params=params, headers={'Content-Type': 'application/json'})
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:156]:
```suggestion
r = requests.put(base_url + "/histories/" + new_hist["id"], json.dumps(data), headers=headers)
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,806 @@
...
+pprint(r.json())
+```
+
+The return value of a PUT request is usually a dictionary with detailed info about the updated resource.
+
+Finally to **delete** a resource, make a DELETE request, e.g.:
+
+```python
+params = {'key': api_key}
+r = requests.delete(base_url + '/histories/' + new_hist['id'], params=params)
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:167]:
```suggestion
r = requests.delete(base_url + "/histories/" + new_hist["id"], headers=headers)
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,806 @@
...
+# Try it out here!
+
+```
+
+
+> ### {% icon question %} Question: New History
+> Next, create a new Galaxy history via POST to the correct API.
+> > ### {% icon solution %} Solution
+> > ```python
+> > params = {'key': api_key}
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:205]:
```suggestion
> > headers = {"Content-Type": "application/json", "x-api-key": api_key}

```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,806 @@
...
+> > data = {'name': 'New history'}
+> > r = requests.post(base_url + '/histories', json.dumps(data), params=params, headers={'Content-Type': 'application/json'})
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:207]:
```suggestion
> > r = requests.post(base_url + "/histories", json.dumps(data), headers=headers)
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,806 @@
...
+> https://raw.githubusercontent.com/nsoranzo/bioblend-tutorial/main/test-data/convert_to_tab.ga
+> ```
+> > ### {% icon solution %} Solution
+> >
+> > ```python
+> > params = {'key': api_key}
+> > with open('test-data/convert_to_tab.ga', 'r') as f:
+> >     workflow_json = json.load(f)
+> > data = {'workflow': workflow_json}
+> > r = requests.post(base_url + '/workflows', json.dumps(data), params=params, headers={'Content-Type': 'application/json'})
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:283]:
```suggestion
> > r = requests.post(base_url + "/workflows", json.dumps(data), headers=headers)
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,806 @@
...
+# Try it out here!
+
+```
+
+> ### {% icon question %} Question: View the WF details
+> View the details of the imported workflow by making a GET request to `/api/workflows`.
+> > ### {% icon solution %} Solution
+> > ```python
+> > params = {'key': api_key}
+> > r = requests.get(base_url + '/workflows/' + wf['id'], params)
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:300]:
```suggestion
> > r = requests.get(base_url + "/workflows/" + wf["id"], headers=headers)
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,806 @@
...
+> **Run** the imported workflow on the uploaded dataset **inside the same history** by making a `POST` request to `/api/workflows/<id>/invocations`. The only needed data are 'history' and 'inputs'.
+> > ### {% icon solution %} Solution
+> >
+> > ```python
+> > params = {'key': api_key}
+> > inputs = {0: {'id': hda['id'], 'src': 'hda'}}
+> > data = {
+> >     'history': 'hist_id=' + new_hist['id'],
+> >     'inputs': inputs}
+> > r = requests.post(base_url + '/workflows/' + wf['id'] + '/invocations', json.dumps(data), params=params, headers={'Content-Type': 'application/json'})
```
Comment (nsoranzo) [inline topics/dev/tutorials/bioblend-api/tutorial.md:324]:
```suggestion
> > r = requests.post(base_url + "/workflows/" + wf["id"] + "/invocations", json.dumps(data), headers=headers)
```

File: topics/dev/tutorials/bioblend-api/tutorial.md [hunk 14]
```diff
@@ -0,0 +1,806 @@
...
+api_key = ''
+base_url = urljoin(server, 'api')
+base_url
+```
+
+We now make a GET request to retrieve all histories owned by a user:
+
+```python
+params = {'key': api_key}
+r = requests.get(base_url + '/histories', params)
```
[the remark this answers is not in the document]
Comment (hexylena) [reply topics/dev/tutorials/bioblend-api/tutorial.md:75]: wow, so many small changes have happened over the years. Glad you caught these.

**Review by shiltemann — APPROVED** [review]

**Comment by shiltemann** [comment]
Thanks @nsoranzo and @hexylena !

**Pull request merged**